provider/openstack/firewaller.go

```
@@ -545,92 +545,139 @@ func (c *neutronFirewaller) setUpGlobalGroup(groupName string, apiPort int) (neu
 // zeroGroup holds the zero security group.
 var zeroGroup neutron.SecurityGroupV2
 
-// ensureGroup returns the security group with name and perms.
+// ensureGroup returns the security group with name and rules.
 // If a group with name does not exist, one will be created.
-// If it exists, its permissions are set to perms.
+// If it exists, its permissions are set to rules.
 func (c *neutronFirewaller) ensureGroup(name string, rules []neutron.RuleInfoV2) (neutron.SecurityGroupV2, error) {
 	neutronClient := c.environ.neutron()
+	var group neutron.SecurityGroupV2
+
 	// First attempt to look up an existing group by name.
 	groupsFound, err := neutronClient.SecurityGroupByNameV2(name)
-	if err == nil {
-		for _, group := range groupsFound {
-			if c.verifyGroupRules(group.Rules, rules) {
-				return group, nil
-			}
+	// a list is returned, but there should be only one
+	if err == nil && len(groupsFound) == 1 {
+		group = groupsFound[0]
+	} else if len(groupsFound) == 0 {
+		// A SecurityGroup with the requested name doesn't exist,
+		// create it.  err will say name not found, so ignore.
+		g, err := neutronClient.CreateSecurityGroupV2(name, "juju group")
+		group = *g
+		if err != nil {
+			return zeroGroup, err
 		}
+	} else {
+		return zeroGroup, errors.Errorf("more than one security group with name: %s", name)
 	}
-	// Doesn't exist, so try and create it.
-	newGroup, err := neutronClient.CreateSecurityGroupV2(name, "juju group")
-	if err != nil {
-		return zeroGroup, err
+
+	have := newRuleInfoSetFromRules(group.Rules)
+	want := newRuleInfoSetFromRuleInfo(rules)
+
+	// Find rules we want to delete, that we have but don't want, and
+	// delete them.
+	remove := make(ruleInfoSet)
+	for k := range have {
+		// Neutron creates 2 egress rules with any new Security Group.
+		// Keep them.
+		if !want[k].found && k.Direction != "egress" {
+			remove[k] = have[k]
+		}
 	}
+	for _, v := range remove {
+		_ = neutronClient.DeleteSecurityGroupRuleV2(v.ruleId)
```

why ignore the error?